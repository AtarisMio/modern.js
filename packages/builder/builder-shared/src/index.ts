<<<<<<< HEAD
export * from './applyDefaultBuilderOptions';
export * from './constants';
export * from './createHook';
export * from './createPluginStore';
export * from './createPublicContext';
export * from './utils';
export * from './fs';
export * from './getBrowserslist';
export * from './logger';
export * from './mergeBuilderConfig';
export * from './onExitProcess';
export * from './pick';
export * from './regexp';
export * from './types';
=======
export * from './types';
export * from './utils';
export * from './logger';
export * from './createHook';
export * from './createPluginStore';
export * from './mergeBuilderConfig';
export * from './createPublicContext';
>>>>>>> ac5486156 (refactor(builder): split builder and provider (#1804))
