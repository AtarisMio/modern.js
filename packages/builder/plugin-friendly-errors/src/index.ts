export { outputPrettyError } from './shared/utils';
export * from './plugin';
