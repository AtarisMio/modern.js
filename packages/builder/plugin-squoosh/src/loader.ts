import { DecoderFinalOptions } from '.';
import type { LoaderDefinition } from 'webpack';
import { ImagePool } from '@squoosh/lib';
import { cpus } from 'os';
import { Buffer } from 'buffer';

const imagePool = new ImagePool(cpus().length - 1);

const imageProcessor = async (source: Buffer, options: DecoderFinalOptions) => {
  const img = imagePool.ingestImage(Buffer.from(source));
  const result = await img.encode({ [options.use]: options });
  return result[options.use];
};

/* eslint-disable @babel/no-invalid-this */
const loader: LoaderDefinition<DecoderFinalOptions> = function loader(source) {
  const callback = this.async();
  const opt = this.getOptions();
  const buf = Buffer.from(source);
  imageProcessor(buf, opt)
    .then(result => callback(null, Buffer.from(result.binary)))
    .catch(err => callback(err));
};
/* eslint-enable */

export default loader;
