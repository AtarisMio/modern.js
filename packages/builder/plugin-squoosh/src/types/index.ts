import { encoders } from '../../compiled/@squoosh/lib';
import {
  AvifEncodeOptions,
  JxlEncodeOptions,
  MozJPEGEncodeOptions,
  OxiPngEncodeOptions,
  WebPEncodeOptions,
  WP2EncodeOptions,
} from '../../compiled/@squoosh/lib/codecs';

export interface DecoderCollection {
  mozjpeg: MozJPEGEncodeOptions;
  webp: WebPEncodeOptions;
  avif: AvifEncodeOptions;
  jxl: JxlEncodeOptions;
  wp2: WP2EncodeOptions;
  oxipng: OxiPngEncodeOptions;
}

export type Decoder = keyof DecoderCollection;

export type Codec = typeof encoders[keyof typeof encoders];
