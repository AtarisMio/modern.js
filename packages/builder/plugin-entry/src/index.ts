import { BuilderPlugin } from '@modern-js/builder-shared';
import type { BuilderPluginAPI } from '@modern-js/builder-webpack-provider';
import path from 'path';
import fs from 'fs-extra';
import type webpack from 'webpack';
import { searchEntry } from './searchEntry';

export interface EntrySearchOptions {
  depth?: number;
  match?: string | string[] | RegExp | RegExp[];
  preferJSX?: boolean;
  exclude?: string | string[] | RegExp | RegExp[];
  preEntry?: string | string[];
  overrideEntrySearch?: true;
  root?: string;
}

export type EntrySearch = webpack.Entry | EntrySearchOptions;

// The plugin compat eden entry search logic
// Port from '@ies/cora-webpack-v5' Entry plugin
export function PluginEntry(
  entryOptions: EntrySearch = {},
): BuilderPlugin<BuilderPluginAPI> {
  return {
    name: 'builder-plugin-entry-search',
    setup(api) {
      api.modifyWebpackChain(async chain => {
        const cwd = api.context.rootPath;
        // ------ Pass on to webpack raw entry ------
        // 1. string
        // 2. array
        // 3. object && overrideEntrySearch isn't set
        if (typeof entryOptions === 'string') {
          chain.entry('index').add(path.resolve(cwd, entryOptions));
          return;
        }

        if (Array.isArray(entryOptions)) {
          entryOptions.forEach(e => {
            chain.entry(path.relative(cwd, e)).add(e);
          });
          return;
        }

        if (
          typeof entryOptions === 'object' &&
          !entryOptions.overrideEntrySearch
        ) {
          // @ts-expect-error webpack entry type
          Object.keys(entryOptions).forEach((e: keyof typeof entryOptions) => {
            if (typeof entryOptions[e] === 'object') {
              chain.entryPoints.set(e, {
                // @ts-expect-error assume input[e] as an object
                values() {
                  return entryOptions[e];
                },
              });
            } else {
              chain.entry(e).add((entryOptions as Record<string, string>)[e]);
            }
          });
          return;
        }

        // ----- Auto search entry -----
        const entrySearchOptions = entryOptions as EntrySearchOptions;
        const { preEntry, root } = entrySearchOptions;
        const srcRoot = root
          ? path.resolve(cwd, root, 'src')
          : path.resolve(cwd, 'src');
        const pagesRoot = root
          ? path.resolve(cwd, root, 'pages')
          : path.resolve(cwd, 'pages');
        const srcPagesRoot = root
          ? path.resolve(cwd, root, 'src', 'pages')
          : path.resolve(cwd, 'src', 'pages');
        let entries: string[] | null = null;
        let entrySearchDir: string | null = null;
        if (await fs.pathExists(srcPagesRoot)) {
          entrySearchDir = srcPagesRoot;
          const searchOptions = Object.assign(entrySearchOptions, {
            depth: Math.max(entrySearchOptions.depth ?? 0, 1),
          });

          entries = await searchEntry(srcPagesRoot, searchOptions);
        }
        if (!entries && (await fs.pathExists(pagesRoot))) {
          entrySearchDir = pagesRoot;
          const searchOptions = Object.assign(entrySearchOptions, {
            depth: Math.max(entrySearchOptions.depth ?? 0, 1),
          });

          entries = await searchEntry(pagesRoot, searchOptions);
        }
        if (!entries && (await fs.pathExists(srcRoot))) {
          entrySearchDir = srcRoot;
          entries = await searchEntry(srcRoot, entrySearchOptions);
        }
        if (!entries && (await fs.pathExists(cwd))) {
          entrySearchDir = cwd;
          entries = await searchEntry(cwd, entrySearchOptions);
        }

        if (!entries || !entries.length) {
          const { chalk } = await import('@modern-js/utils');
          console.error(
            `${chalk.bold.red('Error')}: ` +
              `No entries are found in ${chalk.green(
                entrySearchDir ||
                  `${srcPagesRoot} / ${pagesRoot} / ${srcRoot}${cwd}`,
              )}. `,
          );
          throw new Error('No entries are found.');
        }

        entries.forEach(e => {
          const { dir, name } = path.parse(path.relative(cwd, e));
          const entryPath = path.join(dir, name).replace(/^src[/\\]/, '');

          if (preEntry) {
            const normalizedPreEntry =
              typeof preEntry === 'string' ? [preEntry] : preEntry;

            normalizedPreEntry.forEach(pre => {
              chain.entry(entryPath).add(pre);
            });
          }

          chain.entry(entryPath).add(e);
        });
      });
    },
  };
}
