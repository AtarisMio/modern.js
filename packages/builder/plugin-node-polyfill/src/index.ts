import type { BuilderPlugin } from '@modern-js/builder';
import type { BuilderPluginAPI } from '@modern-js/builder-webpack-provider';
import { isWebTarget } from '@modern-js/builder-shared';
import * as path from 'path';

const getResolveFallback = (nodeLibsBrowser: any) =>
  Object.keys(nodeLibsBrowser).reduce<Record<string, string | false>>(
    (previous, name) => {
      if (nodeLibsBrowser[name]) {
        previous[name] = nodeLibsBrowser[name];
      } else {
        previous[name] = false;
      }
      return previous;
    },
    {},
  );

/**
 * Usage:
 *
 *   const { builderPluginNodePolyfill } = await import('@modern-js/builder-plugin-node-polyfill');
 *
 *   builder.addPlugins([ builderPluginNodePolyfill() ]);
 */
export function builderPluginNodePolyfill(): BuilderPlugin<BuilderPluginAPI> {
  return {
    name: 'builder-plugin-node-polyfill',

    async setup(api) {
      if (api.context.bundlerType === 'rspack') {
        const getPolyfillEntry = () => {
          return path.resolve(
            api.context.cachePath,
            'rspack-node-global-polyfill.js',
          );
        };

        api.onBeforeCreateCompiler(async () => {
          if (isWebTarget(api.context.target)) {
            const fs = await import('@modern-js/utils/fs-extra');
            fs.ensureFileSync(getPolyfillEntry());
            // todo: need toggle?
            fs.writeFileSync(
              getPolyfillEntry(),
              `import { Buffer } from 'buffer';
import { process } from 'process';

globalThis.Buffer = Buffer;
globalThis.process = process;`,
            );
          }
        });

        api.modifyBundlerChain(async (chain, { isServer }) => {
          // it had not need `node polyfill`, if the target is 'node'(server runtime).
          if (isServer) {
            return;
          }

          const { entry } = api.context;

          // global polyfill workaround
          Object.keys(entry).forEach(entryName => {
            chain.entry(entryName).prepend(getPolyfillEntry());
          });

          const { default: nodeLibsBrowser } = await import(
            // @ts-expect-error
            'node-libs-browser'
          );

          chain.resolve.alias.merge(getResolveFallback(nodeLibsBrowser));
        });
        return;
      }

      api.modifyWebpackChain(async (chain, { CHAIN_ID, isServer, webpack }) => {
        // it had not need `node polyfill`, if the target is 'node'(server runtime).
        if (isServer) {
          return;
        }

        const { default: nodeLibsBrowser } = await import(
          // @ts-expect-error
          'node-libs-browser'
        );

        chain
          .plugin(CHAIN_ID.PLUGIN.NODE_POLYFILL_PROVIDE)
          .use(webpack.ProvidePlugin, [
            {
              Buffer: [nodeLibsBrowser.buffer, 'Buffer'],
              process: [nodeLibsBrowser.process],
            },
          ]);

        chain.resolve.fallback.merge(getResolveFallback(nodeLibsBrowser));
      });
    },
  };
}

/**
 * @deprecated Using builderPluginNodePolyfill instead.
 */
export const PluginNodePolyfill = builderPluginNodePolyfill;
