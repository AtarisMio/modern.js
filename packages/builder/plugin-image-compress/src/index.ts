import path from 'path';
import _ from '@modern-js/utils/lodash';
import { BuilderPlugin } from '@modern-js/web-builder';
import { Codecs, FinalOptions, Options } from './types';
import codecs from './shared/codecs';

export const withDefaultOptions = (opt: Options): FinalOptions => {
  const options = _.isString(opt) ? { compress: opt } : opt;
  return _.defaults(options, codecs[options.compress].defaultOptions);
};

export const loaderPath = path.resolve(__dirname, 'loader.js');

export const PluginSquoosh = (...options: Options[]): BuilderPlugin => ({
  name: 'web-builder-plugin-image-compress',
  setup(api) {
    const optsWithDefault = options.length
      ? options
      : (_.keys(codecs) as Codecs[]);
    const opts = optsWithDefault.map(opt => withDefaultOptions(opt));
    api.modifyWebpackChain((chain, { CHAIN_ID }) => {
      _.each(opts, opt => {
        chain.module
          .rule(CHAIN_ID.RULE.IMAGE)
          .use(`web-builder-plugin-image-compress#${opt.compress}`)
          .loader(loaderPath)
          .options(opt);
      });
    });
  },
});
