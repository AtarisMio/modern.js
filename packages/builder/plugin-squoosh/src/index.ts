import path from 'path';
import _ from '@modern-js/utils/lodash';
import { BuilderPlugin } from '@modern-js/web-builder';
import { Decoder, DecoderCollection } from './types';
import { encoders } from '../compiled/@squoosh/lib';

export interface BaseDecoderOptions<T extends Decoder> {
  use: T;
  test?: RegExp;
}

export type ComposedDecoderOptions = Decoder[] | { use: Decoder[] };

export type DecoderOptionsCollection = {
  [K in Decoder]: Partial<DecoderCollection[K]> & BaseDecoderOptions<K>;
};

export type DecoderOptions =
  | ComposedDecoderOptions
  | DecoderOptionsCollection[Decoder];

export type DecoderFinalOptionsCollection = {
  [K in Decoder]: Required<DecoderCollection[K] & BaseDecoderOptions<K>>;
};
export type DecoderFinalOptions = DecoderFinalOptionsCollection[Decoder];

export const compileExtRegExp = (ext: string) => new RegExp(`\\.${ext}$`);

export const withDefaultOptions = (
  opt: DecoderOptionsCollection[Decoder],
): DecoderFinalOptions =>
  _.merge(
    {
      test: compileExtRegExp(encoders[opt.use].extension),
    } as DecoderFinalOptions,
    encoders[opt.use].defaultEncoderOptions,
    opt,
  );

export const loaderPath = path.resolve(__dirname, 'loader.js');

export const normalizeFinalOptions = (
  options: DecoderOptions,
): DecoderFinalOptions[] => {
  if (Array.isArray(options)) {
    return options.map(use => withDefaultOptions({ use }));
  }
  if (Array.isArray(options.use)) {
    return options.use.map(use => withDefaultOptions({ use }));
  }
  if (Object.keys(encoders).includes(options.use)) {
    return [withDefaultOptions(options as DecoderFinalOptions)];
  }
  throw new Error(`Unknown encoder: ${options.use}`);
};

export const compareRegExp = (a: DecoderFinalOptions, b: DecoderFinalOptions) =>
  a.test.toString() === b.test.toString();

export const PluginSquoosh = (...options: DecoderOptions[]): BuilderPlugin => ({
  name: 'web-builder-plugin-squoosh',
  setup(api) {
    const optsWithDefault = options.length
      ? options
      : [_.keys(encoders) as Decoder[]];
    const opts = _(optsWithDefault)
      .map(normalizeFinalOptions)
      .flatten()
      .unionWith(compareRegExp)
      .value();
    api.modifyWebpackChain((chain, { CHAIN_ID }) => {
      _.each(opts, (opt, i) => {
        chain.module
          .rule(CHAIN_ID.RULE.IMAGE)
          .use(`web-builder-plugin-squoosh#${i}`)
          .loader(loaderPath)
          .options(opt);
      });
    });
  },
});
