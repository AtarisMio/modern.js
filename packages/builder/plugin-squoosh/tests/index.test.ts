import { it, expect } from 'vitest';
import { add } from '../src';

it('should take a number and add it to another number', () => {
  expect(add(1, 2)).toBe(3);
});
