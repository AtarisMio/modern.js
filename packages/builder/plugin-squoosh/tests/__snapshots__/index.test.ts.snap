// Vitest Snapshot v1

exports[`normalizeFinalOptions > should handle on array of codec 1`] = `
[
  {
    "chromaDeltaQ": false,
    "cqAlphaLevel": -1,
    "cqLevel": 33,
    "denoiseLevel": 0,
    "sharpness": 0,
    "speed": 6,
    "subsample": 1,
    "test": /\\\\\\.avif\\$/,
    "tileColsLog2": 0,
    "tileRowsLog2": 0,
    "tune": 0,
    "use": "avif",
  },
  {
    "arithmetic": false,
    "auto_subsample": true,
    "baseline": false,
    "chroma_quality": 75,
    "chroma_subsample": 2,
    "color_space": 3,
    "optimize_coding": true,
    "progressive": true,
    "quality": 75,
    "quant_table": 3,
    "separate_chroma_quality": false,
    "smoothing": 0,
    "test": /\\\\\\.jpg\\$/,
    "trellis_loops": 1,
    "trellis_multipass": false,
    "trellis_opt_table": false,
    "trellis_opt_zero": false,
    "use": "mozjpeg",
  },
]
`;

exports[`normalizeFinalOptions > should handle on codec options with ext tester 1`] = `
[
  {
    "arithmetic": false,
    "auto_subsample": true,
    "baseline": false,
    "chroma_quality": 75,
    "chroma_subsample": 2,
    "color_space": 3,
    "optimize_coding": true,
    "progressive": true,
    "quality": 75,
    "quant_table": 3,
    "separate_chroma_quality": false,
    "smoothing": 0,
    "test": /\\\\\\.jpeg\\$/,
    "trellis_loops": 1,
    "trellis_multipass": false,
    "trellis_opt_table": false,
    "trellis_opt_zero": false,
    "use": "mozjpeg",
  },
]
`;

exports[`normalizeFinalOptions > should handle on composed options 1`] = `
[
  {
    "chromaDeltaQ": false,
    "cqAlphaLevel": -1,
    "cqLevel": 33,
    "denoiseLevel": 0,
    "sharpness": 0,
    "speed": 6,
    "subsample": 1,
    "test": /\\\\\\.avif\\$/,
    "tileColsLog2": 0,
    "tileRowsLog2": 0,
    "tune": 0,
    "use": "avif",
  },
  {
    "arithmetic": false,
    "auto_subsample": true,
    "baseline": false,
    "chroma_quality": 75,
    "chroma_subsample": 2,
    "color_space": 3,
    "optimize_coding": true,
    "progressive": true,
    "quality": 75,
    "quant_table": 3,
    "separate_chroma_quality": false,
    "smoothing": 0,
    "test": /\\\\\\.jpg\\$/,
    "trellis_loops": 1,
    "trellis_multipass": false,
    "trellis_opt_table": false,
    "trellis_opt_zero": false,
    "use": "mozjpeg",
  },
]
`;

exports[`normalizeFinalOptions > should handle on empty codec options 1`] = `
[
  {
    "arithmetic": false,
    "auto_subsample": true,
    "baseline": false,
    "chroma_quality": 75,
    "chroma_subsample": 2,
    "color_space": 3,
    "optimize_coding": true,
    "progressive": true,
    "quality": 75,
    "quant_table": 3,
    "separate_chroma_quality": false,
    "smoothing": 0,
    "test": /\\\\\\.jpg\\$/,
    "trellis_loops": 1,
    "trellis_multipass": false,
    "trellis_opt_table": false,
    "trellis_opt_zero": false,
    "use": "mozjpeg",
  },
]
`;

exports[`normalizeFinalOptions > should handle on partial codec options 1`] = `
[
  {
    "arithmetic": false,
    "auto_subsample": true,
    "baseline": false,
    "chroma_quality": 75,
    "chroma_subsample": 2,
    "color_space": 1,
    "optimize_coding": true,
    "progressive": true,
    "quality": 75,
    "quant_table": 3,
    "separate_chroma_quality": false,
    "smoothing": 0,
    "test": /\\\\\\.jpg\\$/,
    "trellis_loops": 1,
    "trellis_multipass": false,
    "trellis_opt_table": false,
    "trellis_opt_zero": false,
    "use": "mozjpeg",
  },
]
`;
